### PR TITLE
Interface for template engines

### PR DESCRIPTION
**Pull request by mfiumara**
Initial draft for having interfaces for template engines. This can reworked and is open for discussion.

applies to #248

### CODE REVIEW

**Comment by ReneWerner87** [comment]
@mfiumara thanks for the effort
the interface looks good, but the usage in the first engine does not

in many cases the engines have additional functionalities, so i think we have to export them as well

for the check if the engine complies with the interface i would suggest we use "embedded interfaces" or "typechecks in test or real code".

**Comment by mfiumara** [comment]
Hi,

Thanks for the feedback I'll come up with a refinement of this draft later tonight by embedding interfaces. I see what you mean and we're indeed not exporting all functions of the engine right now.

**Comment by mfiumara** [comment]
Hey I've added an embedded interface for Pug such that all functions are exported now.

For the remainder of the engines I'd go through all of them in a similar fashion. Let me know if anything else is required before I go through with that. Perhaps I should move the function comments inside the interface definition? Or copy them over?

**Review by efectn — COMMENTED** [review]
I think internal dir is not good idea. Someome else may want to use this interface in their projects. It can be template.go in the project root like how it's on storage repo

**Comment by ReneWerner87** [comment]
> I think internal dir is not good idea. Someome else may want to use this interface in their projects. It can be template.go in the project root like how it's on storage repo 

Yes, good point

**Comment by mfiumara** [comment]
@efectn Moved to root into `template.go`.

**Comment by mfiumara** [comment]
@efectn I've done the following:

- Added struct definition into internal: this should not be used by external packages as far as I understand
- The interface definition remains in the root (template.go)
- Individual engines can now embed the internal struct to have all basic methods promoted into their own interface type

I don't see the commit here yet in the PR, I'm not too familiar with Github but perhaps that takes some time to appear.

**Comment by mfiumara** [comment]
Added second Engine Mustache to illustrate further. I should be able to do the rest fairly quickly let me know if this is the way forward.

**Comment by mfiumara** [comment]
I think I'm done @efectn. Would be great if you can have a look.

**Comment by efectn** [comment]
Can you check hound comments @mfiumara

**Comment by mfiumara** [comment]
@efectn Done

**Comment by efectn** [comment]
@mfiumara can you resolve the conflicts, i removed parse() from engines

**Review by efectn — APPROVED** [review]

**Comment by mfiumara** [comment]
@efectn Rebased and squashed commits.

File: ace/ace.go
Comment (efectn) [inline ace/ace.go:24]: I think we no longer need these

File: slim/slim.go
Comment (efectn) [inline slim/slim.go:61]: We should also add interface to Slim.

**Review by efectn — CHANGES_REQUESTED** [review]

**Review by efectn — APPROVED** [review]

**Comment by efectn** [comment]
> Hi @efectn could this be merged? I had to resolve conflicts again due to other changes being merged.

The changes are OK to me but we should wait for one more review. @ReneWerner87 will review the PR when he is free.

**Comment by ReneWerner87** [comment]
@mfiumara very good work

https://github.com/gofiber/template/pull/251#issuecomment-1543400484
> in many cases the engines have additional functionalities, so i think we have to export them as well

this fact is unfortunately not quite covered

try to explain it again
it can happen that an engine has additional methods beyond the basic functionality, which is totally ok
but currently they are not accessible, because you always return the general interface as return type of the "New" method
it's ok to return the original engine here, because internally it's ensured that it implements the methods of the interface and the users also have the possibility to pass the interface in their tests or methods

**Comment by ReneWerner87** [comment]
By the way, if you made the last change, do we have to count up the major because there is a breaking change included here ?

**Pull request merged**
